src/main/java/org/breedinginsight/brapps/importer/daos/BrAPITrialDAO.java

```diff
@@ -58,109 +71,175 @@ public BrAPITrialDAO(ProgramDAO programDAO, ImportDAO importDAO, BrAPIDAOUtil br
     }
 
-    public List<BrAPITrial> getTrialsByName(List<String> trialNames, Program program) throws ApiException {
-        if(trialNames.isEmpty()) {
-            return Collections.emptyList();
+
+    @PostConstruct
+    public void setup() {
+        // Populate the experiment cache for all programs on startup
+        log.debug("populating experiment cache");
+        List<Program> programs = programDAO.getActive();
+        if (programs != null) {
+            programExperimentCache.populate(programs.stream().map(Program::getId).collect(Collectors.toList()));
         }
+    }
+
+    private Map<String, BrAPITrial> fetchProgramExperiment(UUID programId) throws ApiException {
+        TrialsApi api = brAPIEndpointProvider.get(programDAO.getCoreClient(programId), TrialsApi.class);
 
+        // Get the program
+        List<Program> programs = programDAO.get(programId);
+        if (programs.size() != 1) {
+            throw new InternalServerException("Program was not found for given key");
+        }
+        Program program = programs.get(0);
+
+        // Get the program experiments
         BrAPITrialSearchRequest trialSearch = new BrAPITrialSearchRequest();
-        trialSearch.programDbIds(List.of(program.getBrapiProgram().getProgramDbId()));
-        trialSearch.trialNames(trialNames);
-        TrialsApi api = brAPIEndpointProvider.get(programDAO.getCoreClient(program.getId()), TrialsApi.class);
-        return brAPIDAOUtil.search(
+        trialSearch.externalReferenceIDs(List.of(programId.toString()));
+        trialSearch.externalReferenceSources(
+                List.of(Utilities.generateReferenceSource(referenceSource, ExternalReferenceSource.PROGRAMS))
+        );
+        List<BrAPITrial> programExperiments = brAPIDAOUtil.search(
                 api::searchTrialsPost,
                 api::searchTrialsSearchResultsDbIdGet,
                 trialSearch
         );
+
+        return experimentById(processExperimentsForDisplay(programExperiments, program.getKey()));
+    }
+
+    private Map<String, BrAPITrial> experimentById(List<BrAPITrial> trials) {
+        Map<String, BrAPITrial> experimentById = new HashMap<>();
+        for (BrAPITrial experiment: trials) {
+            BrAPIExternalReference xref = experiment
+                    .getExternalReferences()
+                    .stream()
+                    .filter(reference -> String.format("%s/%s", referenceSource, ExternalReferenceSource.TRIALS).equalsIgnoreCase(reference.getReferenceSource()))
+                    .findFirst().orElseThrow(() -> new IllegalStateException("No BI external reference found"));
+            experimentById.put(xref.getReferenceID(), experiment);
+        }
+        return experimentById;
+    }
+
+    public List<BrAPITrial> getTrialsByName(List<String> trialNames, Program program) throws ApiException {
+        Map<String, BrAPITrial> cache = programExperimentCache.get(program.getId());
+        List<BrAPITrial> trials = new ArrayList<>();
```

Comment:
Tim mentioned that we may eventually use something like [RediSearch](https://redis.io/docs/stack/search/) to optimize these methods, might be worth putting a TODO comment anywhere that will need to be updated. Unless you think it will be obvious.

Reply:
Ok, I added a comment.